src/contrib/tf_op/tvm_dso_op_kernels.cc

```diff
@@ -0,0 +1,285 @@
+class TVMDSOOp : public OpKernel {
+
+private:
+  tvm::runtime::PackedFunc tvm_func;
+  string lib_path;
+  string func_name;
+
+  DataType output_dtype;
+  
+  bool has_static_output_shape;
+  std::vector<int64> static_output_shape;
+
+  void initAttributes(OpKernelConstruction* context) {
+    context->GetAttr("lib_path", &lib_path);
+    context->GetAttr("func_name", &func_name);
+    context->GetAttr("output_dtype", &output_dtype);
+     
+    context->GetAttr("has_static_output_shape", &has_static_output_shape);
+    context->GetAttr("static_output_shape", &static_output_shape);
+  }
+
+ public:
+  explicit TVMDSOOp(OpKernelConstruction* context) : OpKernel(context) {
+
+    // Get attr
+    initAttributes(context);
+
+    // Load TVM function from dynamic library
+    tvm::runtime::Module mod_dylib = tvm::runtime::Module::LoadFromFile(lib_path);
+    LOG(INFO) << "Verify dynamic loading from " << lib_path << " device_type=" << TVMDSOOpTrait<DEVICE_TYPE>::device_type;
+    tvm_func = mod_dylib.GetFunction(func_name);
+    CHECK(tvm_func != nullptr);
+  }
+  
+  void Compute(OpKernelContext* context) override {
+
+    DLTensor args[NUM_INPUTS + 1];
+    TensorAsBuf buf_info[NUM_INPUTS];
+    ShapeContainer shapes[NUM_INPUTS];
+
+    int status;
+    int device_id = TVMDSOOpTrait<DEVICE_TYPE>::device_id(context);
+    int device_type = TVMDSOOpTrait<DEVICE_TYPE>::device_type;
+    
+    DLContext dl_ctx = { DLDeviceType(device_type), device_id };
+
+    // Get output shape
+    TensorShape output_shape;
+    auto& output_shape_tensor = context->input(NUM_INPUTS);
+    if (has_static_output_shape) {
+      // use static output shape
+      const int64* dims = static_output_shape.data();
+      TensorShapeUtils::MakeShape(dims, static_output_shape.size(), &output_shape);
+    } else if (output_shape_tensor.dims() == 1) {
+      // use shape tensor values as output shape
+      const int64* dims = output_shape_tensor.flat<int64>().data();
+      TensorShapeUtils::MakeShape(dims, 1, &output_shape);
+    } else {
+      // use input tensor shape by default
+      output_shape = context->input(0).shape();
+    }
+    
+    for (int i = 0; i < NUM_INPUTS; ++i) {
+        // Grab the input tensor
+        auto& input_tensor = context->input(i);
+
+        // Create shape container, should keep ref during execution
+        shapes[i] = input_tensor.shape().dim_sizes();
+        auto shape_ptr = (int64_t*) shapes[i].data();
+
+        TensorAsBuf& input = buf_info[i];
+        input.device_type = device_type;
+
+        EnsureAlignment(context, input_tensor, &input);
+        input.CopyFromOrigin();
+
+        status = MakeDLTensor(input, dl_ctx, shape_ptr, &args[i]);
+        OP_REQUIRES(context, status == 0, Status(error::INTERNAL, "Fail to create dlpack tensor for input"));
+    }
+
+    // Allocate output tensor
+    Tensor* output_tensor;
+    OP_REQUIRES_OK(context, context->allocate_output(0, output_shape, &output_tensor));
```

Comment:
If it's a static output shape, can we call this in the constructor? We should only need to do this allocation for dynamic output shapes.

Reply:
> If it's a static output shape, can we call this in the constructor? We should only need to do this allocation for dynamic output shapes.

One TF op instance maybe invoked simultaneously unless marked as "stateful".  I think we should take care of thread-safety issues if conduct this optimization